### PR TITLE
initialize blockpool ops, improve VFile

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**

### CODE REVIEW

File: src/kernel/src/fs/file.rs [hunk 1]
```diff
@@ -72,10 +82,15 @@ pub enum VFileType {
...
-    pub write: fn(&VFile, &[u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>,
-    pub ioctl: fn(&VFile, IoCmd, &mut [u8], Option<&VThread>) -> Result<(), Box<dyn Errno>>,
+    pub read: VFileReadFn,
+    pub write: VFileWriteFn,
+    pub ioctl: VFileOpsIoctlFn,
 }
 
+type VFileReadFn = fn(&VFile, &mut [u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>;
+type VFileWriteFn = fn(&VFile, &[u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>;
+type VFileOpsIoctlFn = fn(&VFile, IoCmd, &mut [u8], Option<&VThread>) -> Result<(), Box<dyn Errno>>;
```
Comment (ultimaweapon) [inline src/kernel/src/fs/file.rs:92]: I think suffix with `Fn` is not provide any benefits here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/fs/file.rs [hunk 2]
```diff
@@ -72,10 +82,15 @@ pub enum VFileType {
...
-    pub write: fn(&VFile, &[u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>,
-    pub ioctl: fn(&VFile, IoCmd, &mut [u8], Option<&VThread>) -> Result<(), Box<dyn Errno>>,
+    pub read: VFileRead,
+    pub write: VFileWrite,
+    pub ioctl: VFileOpsIoctl,
 }
 
+type VFileRead = fn(&VFile, &mut [u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>;
+type VFileWrite = fn(&VFile, &[u8], Option<&VThread>) -> Result<usize, Box<dyn Errno>>;
+type VFileOpsIoctl = fn(&VFile, IoCmd, &mut [u8], Option<&VThread>) -> Result<(), Box<dyn Errno>>;
```
Comment (ultimaweapon) [inline src/kernel/src/fs/file.rs:92]: Should be `VFileIoctl` instead of `VFileOpsIoctl`.

File: src/kernel/src/dmem/blockpool.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
+use crate::errno::{Errno, ENOTTY, ENXIO};
+use crate::fs::{IoCmd, VFile, VFileOps};
+use crate::process::VThread;
+use std::num::NonZeroI32;
+use thiserror::Error;
+
+const BLOCKPOOL_FILEOPS: VFileOps = VFileOps {
```
Comment (ultimaweapon) [inline src/kernel/src/dmem/blockpool.rs:7]: Where is this defined in the PS4 kernel?

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: src/kernel/src/dmem/blockpool.rs (hunk 3 above)
Comment (SuchAFuriousDeath) [reply src/kernel/src/dmem/blockpool.rs:7]: look at the finit call in sys_blockpool open

File: src/kernel/src/dmem/blockpool.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    match cmd {
+        BLOCKPOOL_CMD1 => todo!("blockpool ioctl cmd 1"),
+        BLOCKPOOL_CMD2 => todo!("blockpool ioctl cmd 2"),
+        _ => Err(IoctlError::InvalidCommand(cmd).into()),
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum GenericError {
+    #[error("Invalid operation")]
```
Comment (ultimaweapon) [inline src/kernel/src/dmem/blockpool.rs:28]: The message should follow with Rust guidance.

File: src/kernel/src/dmem/blockpool.rs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    fn errno(&self) -> NonZeroI32 {
+        match self {
+            GenericError::InvalidOperation => ENXIO,
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum IoctlError {
+    #[error("Invalid command {0}")]
```
Comment (ultimaweapon) [inline src/kernel/src/dmem/blockpool.rs:42]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**